client/components/common/BaseForm.vue

```diff
@@ -0,0 +1,111 @@
+<!-- Reusable component representing a form that can be used in multiple contexts -->
+
+<template>
+  <article>
+    <span
+      v-for="{id, label, type, options, hint, optional} in fields"
+      :key="id"
+    >
+      <label v-if="type !== 'hidden'" :for="id"> {{ label }}: </label>
+      <small v-if="optional"> (optional) </small>
+      <select
+        v-if="type === 'select'"
+        :class="errors[id] ? 'error' : ''"
+        v-model="values[id]"
+        @change="() => validate(id)"
+      >
+        <option/>
+        <option
+          v-for="option in options"
+          :key="option"
+          :value="option"
+        >
+        {{ option }}
+        </option>
+      </select>
+      <textarea
+        v-else-if="type === 'textarea'"
+        :class="errors[id] ? 'error' : ''"
+        v-model="values[id]"
+        @change="() => validate(id)"
+      />
+      <input
+        v-else
+        :class="errors[id] ? 'error' : ''"
+        :type="type || 'text'"
+        v-model="values[id]"
+        @change="() => validate(id)"
+      >
+      <small v-if="hint"> {{ hint }} </small>
+      <small v-if="errors[id]" class="error"> {{ errors[id] }} </small>
+    </span>
+  </article>
+</template>
+
+<script>
+export default {
+  name: 'BaseForm',
+  props: {
+    fields: {
+      type: Array,
+      required: true
+    },
+    document: {
+      type: Object,
+      required: true
+    },
+    customValidators: {
+      type: Object,
+      required: true
+    }
+  },
+  data() {
+    return {
+      values: Object.assign({}, this.document), // The values of the form
+      validators: {}, // Functions to run for client-side validation
+      errors: {}, // Errors from validators
+    };
+  },
+  created() {
+    // this.values = Object.fromEntries(this.fields.map(f => [f.id, this.document[f.id]]));
+    this.fields.forEach(f => {
+      const validators = [
+        !f.optional && (v => v ? '' : 'required field'),
+        this.customValidators[f.id]
+      ].filter(x => x);
+      this.$set(this.validators, f.id, v => validators.reduce((message, validator) => message || validator(v), null));
+    });
+  },
+  mounted() {
+    this.$emit('interface', {
+      hasErrors: () => {
+        this.fields.forEach(f => this.validate(f.id));
+        return Object.values(this.errors).some(x => x);
+      },
+      values: () => Object.assign({}, this.values),
+      modified: () => {
+        return Object.fromEntries(this.fields.filter(f => this.values[f.id] !== this.document[f.id])
+                                             .map(f => [f.id, this.values[f.id]]));
+      },
+      clear: () => this.values = Object.assign({}, this.document)
+    });
```

Comment:
A little lost though I'm sure it works

Reply:
lost on the individual methods or the interface as a whole? I can add comments if it would be helpful, but basically I am only exposing a few operations to keep the components as modular as possible (rather than using $refs for example).